subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/DynamicVersionResolver.java

```diff
@@ -344,23 +368,24 @@ private void process(ModuleComponentRepositoryAccess access, DefaultBuildableMod
         /**
          * Once a version has been selected, this tries to resolve metadata for this specific version. If it can it
          * will copy the result to the target builder
+         *
          * @param target where to put metadata
+         * @return the resolved metadata, if it could be resoved
          */
-        public void resolve(BuildableModuleComponentMetaDataResolveResult target) {
+        private ModuleComponentResolveMetadata tryResolveMetadata(BuildableModuleComponentMetaDataResolveResult target) {
```

Comment:
I don't like the combination of parameter object plus return value. Why can't we let the parameter object take care of tracking everything we need to know on the caller side?

Reply:
I agree the code is a bit of a mess here. I already simplified it in a previous PR, but it deserves better.

Reply:
Fixed by 59668f6